src/zkp/mod.rs

```diff
@@ -34,15 +41,22 @@ use serde::{de::DeserializeOwned, Serialize};
 pub(crate) trait Proof: Sized + Serialize + DeserializeOwned {
     type CommonInput;
     type ProverSecret;
+    type ProofContext;
```

Comment:
Todo: Remove the associated type here. A proof doesn't have to know exactly what context it expects; it just needs to take _some_ context.

It's fine if that context is different for proving and verifying because verification will just fail.